### PR TITLE
Bug 2052744: kube-apiserver Degraded=True if webhook references kubernetes.default.svc

### PR DESCRIPTION
**Pull request by sanchezl**
We ran into a case where the webhook was implemented in the apiserver, so the webhook configuration referenced the `kubernetes.default.svc` service.
* The webhook configuration did not specify a CABundle to use[1], so the webhook check failed since it attempted to connect without the needed CA certs. The 
* Nevertheless. the webhook should still work due to the [special case in the webhook impl](https://github.com/openshift/kubernetes/blob/9dd59017c4c218135b12949a2bd3af07e892199a/staging/src/k8s.io/apiserver/pkg/util/webhook/authentication.go#L200) that loads the CA certs provided by the service account when accessing the `kubernetes.default.svc` service. 

[1]: It asked for one to be injected, but the apiserver certs are not managed by the service-ca operator that injects CABundles.

### CODE REVIEW

**Comment by aojea** [comment]
can there be some attack angle here?

File: pkg/operator/webhooksupportabilitycontroller/degraded_webhook.go [hunk 1]
```diff
@@ -101,8 +101,15 @@ func (c *webhookSupportabilityController) assertConnect(ctx context.Context, ref
 		port = fmt.Sprintf("%d", *reference.Port)
 	}
 	rootCAs := x509.NewCertPool()
-	if len(caBundle) > 0 {
+	switch {
+	case net.JoinHostPort(host, port) == "kubernetes.default.svc:443":
+		// ignore any caBundle specified and get one from the service account
```
Comment (aojea) [inline pkg/operator/webhooksupportabilitycontroller/degraded_webhook.go:106]: that leaves caBundle field an ignored field for kubernetes.default.svc, right?

Comment (sanchezl) [reply pkg/operator/webhooksupportabilitycontroller/degraded_webhook.go:106]: Correct. This matched the behavior in the apiserver when calling the webhook.

File: pkg/operator/webhooksupportabilitycontroller/webhook_supportability_controller.go [hunk 2]
```diff
@@ -43,6 +46,7 @@ func NewWebhookSupportabilityController(
 		serviceLister:           kubeInformersForAllNamespaces.Core().V1().Services().Lister(),
 		crdLister:               apiExtensionsInformers.Apiextensions().V1().CustomResourceDefinitions().Lister(),
 		clusterVersionLister:    configInformers.Config().V1().ClusterVersions().Lister(),
+		serviceAccountCAs:       defaultCABundle(),
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator/webhooksupportabilitycontroller/webhook_supportability_controller.go]: I removed the caching from the code.

File: pkg/operator/webhooksupportabilitycontroller/degraded_webhook.go [hunk 3]
```diff
@@ -101,8 +102,15 @@ func (c *webhookSupportabilityController) assertConnect(ctx context.Context, ref
...
+		rootCAs = serviceAccountCABundle()
+	case len(caBundle) > 0:
 		rootCAs.AppendCertsFromPEM(caBundle)
+	default:
+		// if no caBundle was specified, use the one from the service account
+		rootCAs = serviceAccountCABundle()
```
Comment (aojea) [inline pkg/operator/webhooksupportabilitycontroller/degraded_webhook.go:113]: talked with @deads2k in slack

> if the host != "kubernetes.default.svc", then serviceaccount/ca.crt should not be used by default.

it seems we should not default it here , then?

**Pull request closed without merging**